### ISSUE TITLE
SchemaPack: MedRSI evidence-bound capability registration

### ISSUE THREAD

**Issue by crates** [entif-ai/rosetta#1634]
## Summary

Create an experimental **MedRSI capability-registration SchemaPack and read-only interoperability profile** for representing the evidence lifecycle of recursively developed clinical tools: committed diagnostic experience, consequence-aware failure priority, capability-gap specification, frozen candidate package, discovery screening, later-cohort trials, registry-context re-evaluation, registration or archival action, and bounded downstream reliance.

The core boundary is simple: a candidate that improves a discovery cohort has earned further testing, not persistent authority. A registered capability has passed one declared policy over particular cohorts and clinical-cost judgments; that still does not establish broad clinical safety, regulatory approval, causal adequacy, or deployment authorization.

This proposal is based on Wu et al.'s MedRSI research and does not imply author endorsement. Rosetta is a free, open-source project, not intended for sale or monetization, and intended to pursue universally applicable advances in AI that could benefit humanity if successfully realized.

## Research basis

- Paper: [MedRSI: Recursive Self-Improvement for Medical Agents via Clinically Aligned Self-Evolution](https://arxiv.org/abs/2609.24838), arXiv:2609.24838 (2026)
- Paper HTML: https://arxiv.org/html/2609.24838
- Linked code repository: https://github.com/ImprintLab/MedRSI
- Entif context: [The Cost of Learning Too Late](https://entif.ai/tags/research/2026/09/06/the-cost-of-learning-too-late/)
- Entif context: [After the Inflection](http://entif.ai/tags/research/2026/09/07/after-the-inflection/)
- Core authority: [Rosetta v3.0.0 Core Spine Specification](https://github.com/entif-ai/rosetta/blob/main/docs/RFCs/Rosetta%20v3.0.0%20Core%20Spine%20Specification.md)
- Receipt contract: https://github.com/entif-ai/rosetta/blob/main/docs/spec/RECEIPT_FAMILY_CONTRACT.md
- Schema authority map: https://github.com/entif-ai/rosetta/blob/main/packages/rosetta-schemas/docs/schema-authority-map.md

MedRSI keeps the reasoning model and general execution workflow fixed while expanding a stable tool registry. Each candidate is packaged with code or weights, schemas, calling description, dependencies, development records, and a version hash. Discovery admission is permissive. Permanent registration requires prospective paired evaluation on later patient cohorts, mean endpoint gain above a prespecified margin, and no increase in clinical cost. The reference registry hash is retained; a candidate must be re-evaluated if the stable registry changes during its trial period.

The paper reports that immediate registration can create a destructive feedback loop: unstable tools contaminate later diagnostic traces, which then misdirect later capability development. In the 30-round ablation, slow registration rejected 43 tools that would have passed an immediate-discovery rule, and 29 later showed negative mean gains. Three trial cohorts reduced the persistent registry relative to one-cohort registration while preserving substantially better long-horizon performance.

Important constraints remain. Consequence scoring depends on reference quality, a frozen judge/rubric, and clinician calibration. The study uses bounded benchmark and private cohorts, fixed protocol settings, selected model backbones, large compute budgets, and retrospective test curves after checkpoint freezing. Prospective clinical studies are still needed. At issue-drafting time, the paper-linked GitHub repository exists but is empty, so implementation must begin from the published protocol and synthetic fixtures unless maintainers later release artifacts under clear terms.

## Why this belongs in Rosetta

The paper supplies a concrete external protocol for a difficult provenance and authority boundary: a self-improving agent changes the evidence-generating environment on which its future improvements depend.

The profile should keep these claims separate:

1. this diagnosis and execution trace were committed before the reference was revealed;
2. this case was judged incorrect;
3. this clinical-consequence score was produced under this rubric, judge, and calibration record;
4. this failure cluster motivated this capability-gap hypothesis;
5. this exact candidate package was built and froze under this source registry;
6. this discovery comparison justified entry to the experimental pool;
7. these later cohorts produced these paired endpoint and clinical-cost deltas;
8. this registry context remained compatible or triggered re-evaluation;
9. this policy registered, kept pending, rejected, replaced, archived, or rolled back the candidate;
10. a downstream system may rely on the resulting capability only within stated scope and caveats.

Existing issues #911 and #948 own general promotion-gate and bounded self-improvement policy. Issues #1594 and #1624 cover other self-improvement research profiles. This issue is narrower: a loss-aware external schema and adapter for consequence-weighted, prospective-cohort capability registration. It consumes those general contracts and must not redefine them.

## Concept and primitive crosswalk

| MedRSI construct | Proposed representation | Mapping status | Required boundary |
|---|---|---|---|
| stable agent H_r=(M,P,T^S_r) | immutable configuration snapshot with model, workflow, registry, and policy refs | direct | configuration identity is not clinical authorization |
| patient batch and partition | governed source/manifest plus cohort artifact | partial | no patient data enters public fixtures |
| committed diagnosis and execution record | Run with ToolCalls and Observations | direct | reference must remain unavailable during execution |
| reference diagnosis | separate Observation with disclosure time and adjudication provenance | direct | reference quality is not assumed |
| error result | Evaluation | direct | correctness and consequence remain separate |
| consequence score and justification | Evaluation with rubric/judge/calibration refs | partial | model judgment is not clinical fact |
| priority q_i=e_i(1+lambda*c_i) | derived Evaluation with formula and parameter version | direct | priority is budget allocation, not severity truth |
| failure cluster | derived artifact with member/evidence refs | direct | clustering does not establish a causal gap |
| capability-gap specification | Conjecture/application artifact | partial | hypothesis remains testable and revisable |
| candidate package | content-addressed artifact manifest | direct | binds code/weights, schemas, dependencies, and training record |
| discovery comparison | paired Evaluation | direct | positive discovery gain grants trial eligibility only |
| experimental pool | application-state projection | direct | pool membership is not stable-registry membership |
| later-cohort trial | paired Run/Evaluation bundle | direct | patient separation, eligibility, and config identity required |
| mean endpoint/cost deltas, K, delta | policy-bound aggregate Evaluations | direct | aggregation policy and missing-cohort behavior preserved |
| reference-registry hash | configuration anchor | direct | stale context forces re-evaluation |
| serial commit | explicit admission Action and successor snapshot | direct | passing Evaluation is not the commit itself |
| reject/archive/pending | explicit state-transition Action | direct | absence must not be interpreted as rejection |
| replacement/predecessor | supersession relation with preserved predecessor | direct | history cannot be overwritten |
| later reliance | scoped Receipt referencing registry and evidence frontier | partial | registration does not confer deployment authority |

## Recommended collaboration artifact

Create a **SchemaPack plus a read-only import adapter and conformance profile**, provisionally named `medrsi-capability-registration-v1alpha1`.

A generic Core change would overfit one research protocol. A full pack that executes MedRSI would imply operational and clinical authority that the current repository does not have. The right first artifact is a namespaced profile that preserves MedRSI-specific records while reusing core Run, Action, ToolCall, Observation, Evaluation, Receipt, and Tapestry semantics.

## Proposed architecture

```text
MedRSI candidate/trial export
  -> local manifest + digest reader
  -> privacy/redaction preflight
  -> deterministic medrsi.* normalization
  -> core lifecycle translation
  -> profile validation + loss report
  -> receipt bundle and closure verification
  -> optional read-only candidate-lifecycle projection
```

- `MedRSIArtifactReader`: local exported records only; no network, model calls, or clinical execution.
- `MedRSIRedactionPreflight`: rejects direct identifiers and undeclared patient-level fields.
- `MedRSINormalizer`: canonicalizes candidate, cohort, policy, rubric, trial, registry, and decision records while retaining source digests.
- `MedRSIProfileValidator`: checks disclosure order, candidate freezing, cohort separation declarations, registry-context compatibility, aggregate derivation, and legal lifecycle transitions.
- `MedRSITranslator`: emits namespaced application artifacts plus core lifecycle tiles.
- `MedRSIReceiptBuilder`: binds source artifacts, Evaluations, admission Actions, successor registry snapshots, and caveats.
- `MedRSILossReport`: reports preserved, normalized, redacted, dropped, unknown, and unavailable fields.
- `MedRSIProjection`: produces a read-only lifecycle view; it cannot register, execute, or authorize a tool.

## Translation and interoperability design

Define schemas for `medrsi.agent_snapshot`, `medrsi.cohort_manifest`, `medrsi.execution_record`, `medrsi.consequence_evaluation`, `medrsi.failure_cluster`, `medrsi.capability_specification`, `medrsi.candidate_manifest`, `medrsi.discovery_evaluation`, `medrsi.trial_evaluation`, `medrsi.registration_policy`, `medrsi.registration_decision`, and `medrsi.loss_report`.

Rules:

- Preserve original path/digest, paper or repository version, adapter/profile version, task/cohort identity, candidate hash, reference-registry hash, and configuration identity.
- Emit committed prediction/execution and later reference disclosure separately with ordering evidence.
- Treat consequence score, justification, missing-information field, rubric, judge, and calibration reference as one bounded Evaluation, never objective harm truth.
- Preserve lambda, K, delta, endpoint and clinical-cost definitions, seed policy, missing-cohort behavior, and aggregation code/version.
- A discovery pass may produce only `experimental_pool_admit`; it cannot produce `stable_registry_commit`.
- Freeze a candidate before trial. Any mutation creates a new identity and evaluation history.
- Keep reference and candidate-augmented runs paired, preserving ordinary planner selection rather than forced tool use.
- Mark an unavailable eligible cohort `pending`; do not silently shorten the trial horizon.
- Require context compatibility immediately before admission. Registry drift produces `reevaluation_required`.
- Represent registered, rejected, pending, archived, superseded, and rolled_back explicitly.
- Preserve rejected candidates and negative trial evidence without allowing their outputs into stable trajectories.
- Treat patient-level data as out of scope for public examples. Synthetic fixtures preserve control-flow semantics without identifiable records.
- Imports are idempotent under canonical source bytes plus adapter/profile version.
- Unknown fields stay in an extension bag and appear in the loss report.

## Repository and implementation impact

Proposed paths, subject to maintainer review:

- `packs/medrsi-capability-registration/pack.json`
- `packs/medrsi-capability-registration/schema/*.schema.json`
- `packs/medrsi-capability-registration/examples/{registered,rejected,pending,registry-drift}/`
- `packs/medrsi-capability-registration/test-vectors/{valid,invalid}/`
- `packages/medrsi-registration-adapter/src/`
- `packages/medrsi-registration-adapter/src/lib/*.spec.ts`
- `packages/medrsi-registration-adapter/README.md`
- `docs/packs/medrsi-capability-registration-profile.md`

Also update pack validation, schema catalog and authority map, semantic audit, receipt-family mapping, package registration, and required handoff/intake records.

### Phase 0: authority, privacy, and fixture lock

- Freeze paper version and crosswalk.
- Confirm upstream artifact formats if code/examples appear.
- Record the currently empty repository and avoid invented file contracts.
- Define no-PHI synthetic fixtures and license boundaries.
- Confirm adjacency with #911, #948, #1594, and #1624.

### Phase 1: schemas and pure normalizer

- Add manifest and twelve schemas.
- Implement canonicalization, digests, stable IDs, extension retention, redaction preflight, and loss reports.
- Validate lifecycle state, candidate immutability, registry linkage, and policy fields.

### Phase 2: lifecycle translation and receipts

- Translate committed runs, later references, correctness/consequence Evaluations, candidate creation, discovery, trials, context checks, and admission Actions.
- Emit successor snapshots and preserve rejected/pending/predecessor artifacts.
- Build and verify receipt closure.

### Phase 3: conformance fixtures

- Add registered, rejected-after-early-gain, pending-for-missing-cohort, and registry-drift fixtures.
- Add negative vectors for leaked references, candidate mutation, overlapping partition declarations, shortened trial horizon, missing judge/rubric identity, discovery-to-stable shortcut, stale-registry admission, hidden clinical-cost increase, and candidate/committed collapse.
- Verify byte-stable re-import and loss reporting.

### Phase 4: researcher review

- Invite author review of candidate freezing, evaluator-copy semantics, consequence evidence, cohort eligibility, dependency packaging, registry drift, and archive semantics.
- Add any explicitly licensed authentic fixture separately from the synthetic privacy boundary.

## Detailed task checklist

- [ ] Pin paper version and source references.
- [ ] Record the linked repository's current empty state and recheck before implementation.
- [ ] Define the synthetic no-PHI fixture policy.
- [ ] Define namespace, manifest, compatibility range, and authority classification.
- [ ] Write the twelve schemas.
- [ ] Implement source digest, canonicalization, stable-ID, extension-bag, and redaction handling.
- [ ] Separate committed diagnosis, reference disclosure, correctness, and consequence artifacts.
- [ ] Preserve judge, prompt, rubric, calibration, and missing-information evidence.
- [ ] Preserve candidate package contents and dependency-chain identity.
- [ ] Enforce candidate freezing and new identity on mutation.
- [ ] Separate discovery admission from persistent registration.
- [ ] Preserve trial eligibility, partition, pairing, and incomplete-history status.
- [ ] Recompute or verify aggregate metrics from referenced evaluations.
- [ ] Enforce registry-context compatibility before admission.
- [ ] Preserve rejection, pending, archive, supersession, and rollback.
- [ ] Emit core lifecycle tiles without redefining core semantics.
- [ ] Emit conformance, receipt-closure, redaction, and loss reports.
- [ ] Add four positive fixtures and at least nine negative vectors.
- [ ] Add deterministic re-import and round-trip tests.
- [ ] Update schema catalog, authority map, semantic audit, pack docs, and handoff/intake records.
- [ ] Run pack/schema, adapter, semantic-governance, and narrow verification tests.

## Acceptance criteria and verification

- [ ] Same source bytes/profile version produce byte-stable normalized output and stable CIDs.
- [ ] Every derived artifact points to original digest and task/cohort/candidate/registry identity.
- [ ] A discovery pass cannot create a stable-registry commit.
- [ ] A mutated candidate cannot inherit prior trial evidence.
- [ ] Reference disclosure cannot precede committed diagnostic trace.
- [ ] Missing cohorts remain pending and cannot be averaged away.
- [ ] Registration requires declared K/delta, positive mean endpoint gain, and non-increasing clinical cost.
- [ ] Registry drift blocks admission until compatible-context evaluation exists.
- [ ] Rejected/archived candidates remain queryable but absent from stable trajectories.
- [ ] Replacement preserves predecessor and supersession lineage.
- [ ] Receipt closure fails with missing candidate, cohort, rubric, evaluation, policy, or registry evidence.
- [ ] Public fixtures contain no real patient data or direct identifiers.
- [ ] Documentation states that conformance/registration do not prove truth, clinical safety, regulatory approval, patient benefit, or deployment authority.

## Risks, constraints, and licensing

- The linked repository is currently empty; file formats, licenses, and implementation details cannot be assumed.
- Private cohorts cannot be reproduced. Use synthetic fixtures unless authors provide de-identified, licensed exports.
- A frozen LLM judge/rubric improves consistency but does not make consequence scoring objective.
- Reference diagnoses and clinician calibration can be wrong or context-bound.
- Mean gains and aggregate clinical cost can conceal subgroup regressions.
- Partition separation is asserted by source manifests unless independently audited.
- Tested tasks, models, budgets, and cohorts do not establish broad medical or RSI generalization.
- Expensive development/evaluation loops complicate replay.
- Current validators are lightweight and headless; this profile must not imply a live clinical gate.
- Copying future upstream code or authentic fixtures requires license review.

## Open questions for the researchers

1. Which export is authoritative for candidate contents, hashes, cohort eligibility, and final registration state?
2. Is consequence justification intended as a stable interface or an internal evaluator trace?
3. How should unresolved scores and later clinician adjudication be represented without overwriting originals?
4. What registry change invalidates accumulated trials: any hash change, dependency change, or demonstrated interaction?
5. Should generator/training-data/predictor chains register atomically or receive separate lifecycle states?
6. Which subgroup checks should be mandatory admission evidence?
7. Can one synthetic lifecycle export be released under an explicit license?
8. How should post-registration monitoring trigger quarantine or rollback?

## Non-goals

- Reimplementing MedRSI, OpenHands, MedAgent-Pro, or a clinical agent.
- Processing real patient data in public fixtures.
- Running training, diagnosis, or consequence judging.
- Defining universal severity or promotion thresholds.
- Replacing #911, #948, #1594, or #1624.
- Granting execution, clinical, regulatory, or deployment authority.
- Claiming a passed rule proves safe recursive self-improvement.
- Adding MedRSI-specific concepts to Core.

## Collaboration path

Start with a maintainer-owned synthetic lifecycle fixture covering one candidate that passes discovery and fails later trials. Implement deterministic translation, redaction preflight, loss reporting, and receipt closure. Then invite the authors to review the crosswalk and, if appropriate, supply an explicitly licensed synthetic export. A second fixture can test registry drift and compound dependencies. Any later connection to live promotion gates remains downstream of #911 and #948.

## Source references

- https://arxiv.org/abs/2609.24838
- https://arxiv.org/html/2609.24838
- https://github.com/ImprintLab/MedRSI
- https://entif.ai/tags/research/2026/09/06/the-cost-of-learning-too-late/
- http://entif.ai/tags/research/2026/09/07/after-the-inflection/
- https://github.com/entif-ai/rosetta/
- https://github.com/entif-ai/rosetta/blob/main/docs/RFCs/Rosetta%20v3.0.0%20Core%20Spine%20Specification.md
- https://github.com/entif-ai/rosetta/blob/main/docs/spec/RECEIPT_FAMILY_CONTRACT.md
- https://github.com/entif-ai/rosetta/blob/main/packages/rosetta-schemas/docs/schema-authority-map.md
- Related issues: #911, #948, #1594, #1624
